portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -1,143 +1,159 @@
+  } 
+
+  /**
+   * Sets up strong consistency for datastore service. 
+   * This strong consistency ensures that freshness is more important than availability
+   * so that the most up-to-date data is returned and displayed on the page.
+   */ 
+  private DatastoreService getDatastoreServiceWithConsistency() { 
+    DatastoreServiceConfig datastoreConfig = 
+        DatastoreServiceConfig.Builder.withReadPolicy(new ReadPolicy(Consistency.STRONG)).deadline(5.0);
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService(datastoreConfig);
+    return datastore; 
   }
 
   /**
-   * Receives the quote that the user has inputed in the survey bar.
-   * This new quote is changed into an Entity that is put into Datastore.
-   * Each quote entity has a "text" and a "timstamp" field. 
+   * Takes in an ArrayList of quotes and converts it into a string 
+   * that is in JSON format. 
+   * @param a list of quotes
+   * @return the quote list formated as a JSON string
    */ 
-  @Override
-  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {    
-    String quote = request.getParameter("quote");
+  private String convertQuotesToJson(List<Quote> quoteList) {  
+    Gson gson = new Gson();
+    String json = gson.toJson(quoteList);
+    return json;
+  }
 
-    // Only update the new quote if it is not an empty string. 
+  /**
+   * Creates a new entity for each quote and puts each new quote into Datastore
+   * @param request: the HTTP request; 
+   *        userInfo: a HashMap that contains user information connected to each quote
+   */ 
+  private void putQuoteIntoDatastore(HttpServletRequest request, User user) {
+    String quote = request.getParameter("quote");   
     if (quote.length() > 0) {
       long timestampMillis = System.currentTimeMillis();
 
       Entity quoteEntity = new Entity(QUOTE);
       quoteEntity.setProperty(TEXT, quote);
       quoteEntity.setProperty(TIMESTAMP, timestampMillis);
+      quoteEntity.setProperty(USER_EMAIL, user.getEmail()); 
+      quoteEntity.setProperty(NICKNAME, getUserNickname(user.getUserId()));
 
-      DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+      DatastoreService datastore = getDatastoreServiceWithConsistency();
       datastore.put(quoteEntity);
     }
-    
-    response.sendRedirect("/index.html");
   }
 
   /**
-   * Extracts the attributes of a quoteEntity and puts them into a new
-   * instance of Quote. 
+   * Fetches quotes from Datastore and puts them into a list in JSON format. 
    */ 
-  private Quote extractQuoteFromEntity(Entity quoteEntity) {
-    long id = quoteEntity.getKey().getId();
-    String text = (String) quoteEntity.getProperty(TEXT);
-    long timestampMillis = (long) quoteEntity.getProperty(TIMESTAMP);
-    Quote newQuote = new Quote(id, text, timestampMillis);
-    return newQuote; 
-  }
+  private String getQuoteListJson(HttpServletRequest request) {
```

Comment:
It's generally a good practice to pass in the minimal set of arguments that you can, and I think in this case if we call `getNumQuoteDisplayed(...)` before this method then you can just pass in that count here instead of the whole request.

Passing in the minimal data makes it easier to reason about the method from the outside because there are just fewer things that the method could be doing, since it has access to less data.

Reply:
I see!